File: python/pia/resources/news.py
"""Official PIA SDK - News Intelligence Resource."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from ..transport import AsyncTransport, SyncTransport
from ..types import NewsFeedResponse


class NewsResource:
    """Synchronous News API resource."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def get_news(
        self,
        *,
        symbols: Optional[List[str]] = None,
        category: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> NewsFeedResponse:
        """Fetches curated financial news articles with ticker entity tagging."""
        params: Dict[str, Any] = {}
        if symbols:
            params["symbols"] = ",".join(s.strip().upper() for s in symbols if s.strip())
        if category:
            params["category"] = category
        if limit is not None:
            params["limit"] = limit

        data = self._transport.request("/api/v1/news", method="GET", params=params)
        return NewsFeedResponse.from_dict(data)

    def get_latest(self) -> Dict[str, Any]:
        """Fetches latest breaking news."""
        return self._transport.request("/api/v1/news/latest", method="GET")

    def get_by_id(self, article_id: str) -> Dict[str, Any]:
        """Fetches specific news article by ID."""
        return self._transport.request(f"/api/v1/news/{article_id}", method="GET")


class AsyncNewsResource:
    """Asynchronous News API resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def get_news(
        self,
        *,
        symbols: Optional[List[str]] = None,
        category: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> NewsFeedResponse:
        """Fetches curated financial news articles with ticker entity tagging."""
        params: Dict[str, Any] = {}
        if symbols:
            params["symbols"] = ",".join(s.strip().upper() for s in symbols if s.strip())
        if category:
            params["category"] = category
        if limit is not None:
            params["limit"] = limit

        data = await self._transport.request("/api/v1/news", method="GET", params=params)
        return NewsFeedResponse.from_dict(data)

    async def get_latest(self) -> Dict[str, Any]:
        """Fetches latest breaking news."""
        return await self._transport.request("/api/v1/news/latest", method="GET")

    async def get_by_id(self, article_id: str) -> Dict[str, Any]:
        """Fetches specific news article by ID."""
        return await self._transport.request(f"/api/v1/news/{article_id}", method="GET")

File: typescript/src/resources/news.ts
/**
 * Official PIA SDK - News Intelligence API Resource
 */

import type { HttpTransport } from "../http/transport";
import type { GetNewsOptions, NewsFeedResponse, RequestOptions } from "../types";

export class NewsResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Fetches curated financial and macro news articles.
   */
  public async getNews(options?: GetNewsOptions): Promise<NewsFeedResponse> {
    const params = new URLSearchParams();
    if (options?.symbols && options.symbols.length > 0) {
      params.set("symbols", options.symbols.map((s) => s.trim().toUpperCase()).join(","));
    }
    if (options?.category) params.set("category", options.category);
    if (options?.limit) params.set("limit", String(options.limit));

    const query = params.toString() ? `?${params.toString()}` : "";
    return this.transport.request<NewsFeedResponse>(
      `/api/v1/news${query}`,
      "GET",
      undefined,
      options
    );
  }

  /**
   * Fetches the latest breaking financial news bulletin.
   */
  public async getLatest(options?: RequestOptions): Promise<any> {
    return this.transport.request<any>(
      "/api/v1/news/latest",
      "GET",
      undefined,
      options
    );
  }

  /**
   * Fetches detailed intelligence for a specific news article ID.
   */
  public async getById(id: string, options?: RequestOptions): Promise<any> {
    return this.transport.request<any>(
      `/api/v1/news/${encodeURIComponent(id)}`,
      "GET",
      undefined,
      options
    );
  }
}
